pkg/registry/apps/deployment/strategy.go

```diff
@@ -73,12 +73,15 @@ func (deploymentStrategy) GetResetFields() map[fieldpath.APIVersion]*fieldpath.S
 	return fields
 }
 
-// PrepareForCreate clears fields that are not allowed to be set by end users on creation.
+// PrepareForCreate clears fields that are not allowed to be set by end users on creation
+// and truncate deployment name
 func (deploymentStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
 	deployment := obj.(*apps.Deployment)
 	deployment.Status = apps.DeploymentStatus{}
 	deployment.Generation = 1
 
+	deployment.SetName(TruncateDeploymentName(deployment.GetName()))
```

Comment:
You definitely cannot do this, it could cause separate manifests that work today to collide on name in the next version.

Additionally it doesn't fix already existing deployments.

Reply:
If truncate deployment name when its length longer than 253,its that ok?because a deployment with name length longer than 253 will failed to create rs. Or can you give some advice. thx.

Reply:
uhhhh for things that *currently don't work* I'd have to think about it, I think this would still be a last resort and instead warning + validation change would be preferred

Reply:
We definitely cannot mutate the name of the object being created like this. That breaks all kinds of things like server-side apply creation being idempotent, the object returned having a different name than the object submitted, authorizing create of deployment xyz actually allowing create of deployment xy, etc

Reply:
If deployments with long names don't work today because they produce downstream replicasets with names that are too long, the way to fix that is to change the shape of the replicasets that are produced from deployments with long names, not to mutate the long deployment names

Reply:
Agree - we can truncate the name when creating the RS because there's no real contract there.  We _cannot_ truncate the Deployment's own name - that is user-input and we don't mutate explicit user input.

Was there a SIG discussion on this topic?  https://github.com/kubernetes/kubernetes/pull/116830#issuecomment-1481492252

Reply:
> Was there a SIG discussion on this topic? [#116830 (comment)]

Had a brief discussion with lavalamp on slack.Or the SIG discussion mean make a discussion on mailing list or some where(sorry for this question)?
> Agree - we can truncate the name when creating the RS because there's no real contract there
This is my mistake,misunderstading the option 3. I will modify this.

Reply:
Probably the best venue is the SIG Apps meeting https://github.com/kubernetes/community/tree/master/sig-apps

Reply:
Yeah, the SIG meeting or mailing list or slack all could work. My point was that the people is sig-apps have better context than I do or probably lavalamp does, and would be more likely to know if this will cause some downstream breakage.

Reply:
Ok, I got it.  thinks.

Reply:
@thockin I only get one reply about this issue on slack. But SIG apps have meetings every two weeks ansd April 17 meeting was cancelled due to KubeCon EU. So determine which option to apply would be after April 24.

Reply:
Yeah. I will write a KEP. About feature gate i may be need to learn how to do that.

Reply:
Don't worry!  Feature gates are easy.  It's just a flag that you can check to enable (or not) your new logic.

Reply:
> Probably the best venue is the SIG Apps meeting https://github.com/kubernetes/community/tree/master/sig-apps

@alculquicondor Should I edit [agenda](https://docs.google.com/document/d/1LZLBGW2wRDwAfdBNHJjFfk9CFoyZPcIYGWU7R1PQ3ng/edit#) before SIG Apps meeting? But i don't have edit permission . And i applied for permission a few days ago, but no response.

Reply:
You can join kubernetes-sig-apps@googlegroups.com to gain permissions